src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs

```diff
@@ -98,7 +104,11 @@ protected override Task DisposeCoreAsync(bool initialized)
             // active configuration should be updated before resetting subscriptions
             await RefreshActiveConfigurationAsync().ConfigureAwait(false);
 
-            _designTimeBuildSubscriptionLink?.Dispose();
+            foreach (var disposable in _designTimeBuildSubscriptionLinks)
+            {
+                disposable.Dispose();
+            }
+            _designTimeBuildSubscriptionLinks.Clear();
```

Comment:
`OnceInitializedOnceDisposed` only makes the initialize and disposal mutually exclusive. It does not guard against any other code in your class being executed during disposal or initialization.

Not only can you have the the list change during enumeration, you could also dispose and _then_ recreate the entire list, which will then never be disposed. Our `DisposableBag` handles that, anything added after disposal is immediately disposed. Its a shame it isn't public. You can copy it out though. 

You either need to add synchronization logic around the list and check if you are disposed before adding new links to the list, or use a data structure like `DisposableBag` that handles that for youl